### ISSUE TITLE
Measurement: Add snapping to measurements / picking points

### ISSUE THREAD

**Issue by maxwxyz** [FreeCAD/FreeCAD#13708]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Currently when selecting two circles or cylinders, the measurement is displayed between the enveloped circle curve or between the mantle surfaces.
It would be great if the unified measurement facility could support snapping like in the Draft WB to snap to circle midpoints, intersections, and so on.
This would also allow to measure distances and angles of axes between shafts and holes.
![grafik](https://github.com/FreeCAD/FreeCAD/assets/6246609/65ba0d04-a7de-477c-af6a-cb272345fb40)

### Full version info

```shell
OS: Windows 10 build 19045
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37063 (Git)
Build type: Release
Branch: main
Hash: 680792030fd664c94829e2c1b78b21788e8d9879
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: German/Germany (de_DE)
Installed mods: 
  * BIM 2021.12.0
  * CurvedShapes 1.0.9
  * Curves 0.6.35
  * fasteners 0.5.20
  * OpenTheme 2024.4.20
  * OpticsWorkbench 1.0.17
  * sheetmetal 0.4.13

```

### Subproject(s) affected?

Core

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by obelisk79** [comment]
@hlorus is there a chance you could get a PR submitted for this soon? It's important enough that I think there is an adequate argument to have this merged despite the feature freeze. The measure tool in the Manipulator addon is a good example of how this could work.

**Comment by FEA-eng** [comment]
I'll just add that, based on feedback from users in various FreeCAD communities, this is clearly the most important missing feature of the new Measure tool. Particularly measuring using centers of holes instead of their edges. Some users even thought it was the primary design goal for this tool.

**Comment by maxwxyz** [comment]
@hlorus any chance? @Roy-043 is it possible to use Draft snapping functions?

**Comment by Roy-043** [comment]
Draft snaps currently cannot be used outside the Draft WB.

**Comment by hlorus** [comment]
A quick update here: I did have a look into this and have a WIP branch. Since measurements can be persistent simply snapping to some positions doesn't make sense because the measurement wouldn't follow the model after any modification is made. Therefor i propose to use PartDesign's datum elements for this and add functionality to the measure command to implicitly create them when a modifier key is pressed. 
@wwmayer does this sound like a reasonable solution?

**Comment by wwmayer** [comment]
> Since measurements can be persistent simply snapping to some positions doesn't make sense because the measurement wouldn't follow the model after any modification is made.

I don't think this request is about to click on an arbitrary point of a model and use this point for the measurement. The tool still references the clicked sub-shape of the model so that it remains parametric.

> Therefor i propose to use PartDesign's datum elements for this and add functionality to the measure command to implicitly create them when a modifier key is pressed.

I can't follow that. What do you want to achieve with datum elements?

**Comment by wwmayer** [comment]
> Currently when selecting two circles or cylinders, the measurement is displayed between the enveloped circle curve or between the mantle surfaces.

When talking about two cylinders what are the prerequisites for this? Do they have to be parallel? And are they assumed to be infinite so that you get the distance between two infinite parallel lines or must the finiteness of the cylinders be considered?

**Comment by hlorus** [comment]
> I can't follow that. What do you want to achieve with datum elements?

Allow the user to measure geometry that isn't actually existing. I might be getting this wrong but the elememts described in the issue (circle center, axis of cylinder, midpoint) are not actually existing. So as a workaround the user could create datum elements and measure between them. What i propose is to auto create these datum elements such that the user doesn'g have to manually create them.

As an example: when the user selects a line while pressing the modofier key the command implicitly creates a DatumPoint which is linked to that line and patches the selevtion with the created datum.

**Comment by hlorus** [comment]
> > I can't follow that. What do you want to achieve with datum elements?
> 
> Allow the user to measure geometry that isn't actually existing. I might be getting this wrong but the elememts described in the issue (circle center, axis of cylinder, midpoint) are not actually existing, right? So as a workaround the user could create datum elements and measure between them. What i propose is to auto create these datum elements such that the user doesn'g have to manually create them.
> 
> As an example: when the user selects a line while pressing the modofier key the command implicitly creates a DatumPoint which is linked to that line and patches the selevtion with the created datum.
> 
>

**Comment by wwmayer** [comment]
> Allow the user to measure geometry that isn't actually existing. I might be getting this wrong but the elememts described in the issue (circle center, axis of cylinder, midpoint) are not actually existing, right?

That's right. But I don't think using datum elements will be a way to go because this can become very complicated from programming point of view and for users it adds complexity to the tree view because a datum element requires a body as parent object.

A simple Part::Vertex would do the job, too.

But the problem will be to keep things parametric because if the original shape changes but the created vertex (or datum point) isn't adjusted then nothing is won.

**Comment by maxwxyz** [comment]
Adding from https://github.com/FreeCAD/FreeCAD/pull/17416 so it won't get lost: Snapping to circle midpoint was now somehow integrated as default behavior without an option to disable the snapping. If snapping is to be implemented, this behavior should be reverted.

In this case I wanted to measure the distance between the hole and the outer edge:
![grafik](https://github.com/user-attachments/assets/e321d3c5-503a-440e-aede-90067805434f)

And the result was:
![grafik](https://github.com/user-attachments/assets/7e66d1af-3239-4017-a726-43ce337b4df2)

**Comment by eblanshey** [comment]
Would it be possible to add options to display the minimum and maximum distance between the selected elements? In @maxwxyz's example above of "In this case I wanted to measure the distance between the hole and the outer edge", selecting "Minimum distance" would automatically calculate the shortest possible distance between the selected elements, which would be the _edge_ of the hole, and the nearest point on the curved line. In this example a "maximum" distance wouldn't be as useful, but would theoretically select the distance on the opposite side of the hole and the very end of the curved line.

**Comment by ErtugrulEge** [comment]
In this situation you can't measure some simple drawings with measurement tool.Draw a simple rectange and diaogonals and try to measure half of the diagonal.You can't you have to use draft dimension tool.Just annother example

<img width="1378" height="752" alt="Image" src="https://github.com/user-attachments/assets/e753a083-fd19-46e2-9b7f-7e58bc6dc55d" />

```
OS: Windows 10 build 19045
Architecture: x86_64
Version: 1.1.0dev.14555 (Git shallow)
Build date: 2025/09/16 20:38:23
Build type: Release
Branch: main
Hash: bac49e465908071e7400a547eb813bc5017ae7dc
Python 3.11.13, Qt 6.8.3, Coin 4.0.3, Vtk 9.3.1, boost 1_86, Eigen3 3.4.0, PySide 6.8.3
shiboken 6.8.3, xerces-c 3.3.0, IfcOpenShell 0.8.2, OCC 7.8.1
Locale: Turkish/Türkiye (tr_TR)
Navigation Style/Orbit Style/Rotation Mode: CAD/Free Turntable/Drag at cursor
Stylesheet/Theme/QtStyle: unset/FreeCAD Classic/fusion
Logical DPI/Physical DPI/Pixel Ratio: 96/127.628/1
```

**Comment by devanshibhatt23** [comment]
> Of course, these are an important part, but I think a polished approach to the basic snapping would be much preferable from the start.

Okay, I will keep that in mind. Thank you.

**Comment by ayyaz20003** [comment]
Hi everyone,

I'm Ayyaz Sohail, working on a GSoC 2026 proposal for this. I've been digging through the Measure module code and wanted to share a thought on the parametric persistence problem that @wwmayer raised.

The tricky part is that measurements store geometry refs in PropertyLinkSub (like "Edge3"), and snap points (circle centers, midpoints) need to survive model changes — so raw 3D coordinates won't cut it. Datum elements were proposed but rejected as too complex.

Idea: encode snap type as a suffix on the subelement name.

So instead of "Edge3", you'd store "Edge3@center" or "Edge1@midpoint". The @ character isn't used in OCCT naming or FreeCAD's element map system, and PropertyLinkSub already accepts arbitrary strings — no schema changes needed.

On execute(), the measurement strips @center, gets the current Edge3 shape, and re-derives the center via BRepAdaptor_Curve → gp_Circ::Location() — same pattern already in MeasureRadiusHandler. Model changes? Edge3 updates, center gets re-derived. Fully parametric, zero extra objects.

Bonus: a snap-resolved "Edge3@center" could report as MeasureElementType::POINT instead of CIRCLE, unlocking new combos like hole-center-to-vertex distance.

On the Gui side, TaskMeasure::onSelectionChanged() currently skips SetPreselect events. A snap manager could hook in there to detect candidates on hover, show a Coin3D marker, and append the @suffix on click.

Would love feedback on whether @-suffixed subelement names in PropertyLinkSub would be acceptable. Happy to discuss further!
